lib/internal/modules/esm/worker.js

```diff
@@ -0,0 +1,225 @@
+const assert = require('internal/assert');
+const { clearImmediate, setImmediate } = require('timers');
+const {
+  hasUncaughtExceptionCaptureCallback,
+} = require('internal/process/execution');
+const {
+  isArrayBuffer,
+  isDataView,
+  isTypedArray,
+} = require('util/types');
+const { write } = require('fs');
+
+const { receiveMessageOnPort } = require('internal/worker/io');
+const {
+  IDLENESS_OF_THREADS,
+  MAIN_THREAD_IS_IDLE,
+  NUMBER_OF_INCOMING_ASYNC_RESPONSES,
+  NUMBER_OF_MESSAGES_IN_TRANSIT,
+  WORKER_THREAD_IS_IDLE,
+  WORKER_TO_MAIN_THREAD_NOTIFICATION,
+} = require('internal/modules/esm/shared_constants');
+const { initializeESM, initializeHooks } = require('internal/modules/esm/utils');
+
+function transferArrayBuffer(hasError, source) {
+  if (hasError || source == null) return;
+  if (isArrayBuffer(source)) return [source];
+  if (isTypedArray(source)) return [TypedArrayPrototypeGetBuffer(source)];
+  if (isDataView(source)) return [DataViewPrototypeGetBuffer(source)];
+}
+
+function wrapMessage(status, body) {
+  if (status === 'success' || body === null || typeof body !== 'object') {
+    return { status, body };
+  }
+  if (body == null) {
+    return { status, body };
+  }
+  let serialized;
+  let serializationFailed;
+  try {
+    const { serializeError } = require('internal/error_serdes');
+    serialized = serializeError(body);
+  } catch {
+    serializationFailed = true;
+  }
+  return {
+    status,
+    body: {
+      serialized,
+      serializationFailed,
+    },
+  };
+}
+
+async function customizedModuleWorker(lock, syncCommPort, errorHandler) {
+  let hooks, preloadScripts, initializationError;
+  let hasInitializationError = false;
+
+  // DEBUG: let's confirm that using a write operation would let the test pass instead of fail.
+  process.stdout._writev = function _writev(chunks, cb) {
+    const { chunk, encoding } = ArrayPrototypePop(chunks);
+    write(1, chunk, { __proto__: null }, encoding, (err) => {
+      if (err) cb(err);
+      else if (chunks.length === 0) cb();
+      else this._writev(chunks, cb);
+    });
+  };
+
+  try {
+    initializeESM();
+    const initResult = await initializeHooks();
+    hooks = initResult.hooks;
+    preloadScripts = initResult.preloadScripts;
+  } catch (exception) {
+    // If there was an error while parsing and executing a user loader, for
+    // example if because a loader contained a syntax error, then we need to
+    // send the error to the main thread so it can be thrown and printed.
+    hasInitializationError = true;
+    initializationError = exception;
+  }
+
+  syncCommPort.on('message', handleMessage);
+
+  if (hasInitializationError) {
+    syncCommPort.postMessage(wrapMessage('error', initializationError));
+  } else {
+    syncCommPort.postMessage(wrapMessage('success', { preloadScripts }), preloadScripts.map(({ port }) => port));
+  }
+
+  AtomicsNotify(lock, WORKER_TO_MAIN_THREAD_NOTIFICATION); // Unlock the main thread now that we are ready.
```

Comment:
so since atomics can apparently be disabled via a flag, what happens here when it's disabled?

Reply:
It throws `undefined is not a function` or something like this (in reality, it wouldn't arrive here at all, because the destructuring of `Atomics` at the top-level of module have thrown long before the code can get there)

Reply:
and we want an internal module to implicitly throw on evaluation when node is invoked with a flag?

Reply:
I personally don't have a problem with that, there are quite a lot of V8 flags that do not work with Node.js. Using this flag is not supported, if you want Node.js to work, do not use that flag :)

Reply:
lol fair enough. in that case, why not add the atomics builtins to primordials?

Reply:
We can't, because it can be disabled with a runtime flag (it's doesn't exist until after `primordials` is frozen).

Reply:
ahhh gotcha. thanks, feel free to resolve this one.